ewatercycle/config/_validated_config.py

```diff
@@ -0,0 +1,91 @@
+"""Config validation objects."""
+
+import pprint
+import warnings
+from collections.abc import MutableMapping
+from typing import Callable, Dict, Tuple
+import re
+
+from ._validators import ValidationError
+
+
+class InvalidConfigParameter(Exception):
+    """Config parameter is invalid."""
+
+
+class MissingConfigParameter(UserWarning):
+    """Config parameter is missing."""
+
+
+# The code for this class was take from matplotlib (v3.3) and modified to
+# fit the needs of eWaterCycle. Matplotlib is licenced under the terms of
+# the the 'Python Software Foundation License'
+# (https://www.python.org/psf/license)
+class ValidatedConfig(MutableMapping):
```

Comment:
I tried to feed my own config and got unexpected `InvalidConfigParameter: `container_engine` is not a valid config parameter.` exception

mine.yml
```yaml
esmvaltool_config: ~/.esmvaltool/config-user.yml
grdc_location: /data/grdc
container_engine: singularity
singularity_dir: /data/singularity-images
output_dir: /scratch
# Created with cd  /data/singularity-images && singularity pull docker://ewatercycle/wflow-grpc4bmi:2020.1.1
wflow.singularity_images: wflow-grpc4bmi_2020.1.1.sif
wflow.docker_images: ewatercycle/wflow-grpc4bmi:2020.1.1
```

```python
In [1]: from ewatercycle.config import Config
In [2]: c = Config._load_user_config('mine.yml')
In [3]: c
Out[3]: 
Config({'container_engine': 'singularity',
        'esmvaltool_config': PosixPath('/home/verhoes/.esmvaltool/config-user.yml'),
        'ewatercycle_config': PosixPath('/home/verhoes/git/eWaterCycle/ewatercycle/mine.yml'),
        'grdc_location': PosixPath('/data/grdc'),
        'output_dir': PosixPath('/scratch'),
        'singularity_dir': PosixPath('/data/singularity-images'),
        'wflow.docker_images': 'ewatercycle/wflow-grpc4bmi:2020.1.1',
        'wflow.singularity_images': 'wflow-grpc4bmi_2020.1.1.sif'})

In [5]: from ewatercycle.config._validated_config import ValidatedConfig

In [6]: ValidatedConfig(c)
---------------------------------------------------------------------------
InvalidConfigParameter                    Traceback (most recent call last)
<ipython-input-6-d3b3454660ac> in <module>
----> 1 ValidatedConfig(Config._load_user_config('mine.yml'))

~/git/eWaterCycle/ewatercycle/ewatercycle/config/_validated_config.py in __init__(self, *args, **kwargs)
     31         super().__init__()
     32         self._mapping = {}
---> 33         self.update(*args, **kwargs)
     34 
     35     def __setitem__(self, key, val):

~/miniconda38/envs/ewatercycle/lib/python3.9/_collections_abc.py in update(self, other, **kwds)
    938         if isinstance(other, Mapping):
    939             for key in other:
--> 940                 self[key] = other[key]
    941         elif hasattr(other, "keys"):
    942             for key in other.keys():

~/git/eWaterCycle/ewatercycle/ewatercycle/config/_validated_config.py in __setitem__(self, key, val)
     40             raise InvalidConfigParameter(f"Key `{key}`: {verr}") from None
     41         except KeyError:
---> 42             raise InvalidConfigParameter(
     43                 f"`{key}` is not a valid config parameter.") from None
     44 

InvalidConfigParameter: `container_engine` is not a valid config parameter.

```

Am I using it wrong?

Reply:
I expected errors from the non existing paths.

Reply:
You are using it correctly, in the sense that this works as expected:

```python
In [1]: from ewatercycle.config import Config
In [2]: c = Config._load_user_config('mine.yml')
In [3]: c
Out[3]: 
Config({'container_engine': 'singularity',
        'esmvaltool_config': PosixPath('/home/verhoes/.esmvaltool/config-user.yml'),
        'ewatercycle_config': PosixPath('/home/verhoes/git/eWaterCycle/ewatercycle/mine.yml'),
        'grdc_location': PosixPath('/data/grdc'),
        'output_dir': PosixPath('/scratch'),
        'singularity_dir': PosixPath('/data/singularity-images'),
        'wflow.docker_images': 'ewatercycle/wflow-grpc4bmi:2020.1.1',
        'wflow.singularity_images': 'wflow-grpc4bmi_2020.1.1.sif'})
```

Note how the paths are loaded as `pathlib.PosixPath` objects. So it validates the type, but it does not check if the paths exist. We discussed this briefly, and we agreed that this is not necessary. If we need the validator to check whether the paths exist, it can be added.

```bash
ValidatedConfig(c)
```
This will not work, because `ValidatedConfig` is the base class that `Config` inherits from. It does not know which config parameters are associated with the config. `Config` has an attribute `._validate`, which is a dict of validators to apply to every key/value pair. This one is absent in `ValidatedConfig`.

Reply:
Ok, thanks so paths are not checked for existence. Having typos in keys, missing keys all produce expected warnings and values.